shared/src/main/scala/mlscript/NormalForms.scala

```diff
@@ -53,6 +68,9 @@ class NormalForms extends TyperDatatypes { self: Typer =>
           case (S(b), w: Without) => S(Without(b & w, SortedSet.empty)(noProv))
           case (S(w: Without), b) => S(Without(w & b, SortedSet.empty)(noProv))
           case (S(_), _) => N
+          case (N, tup: TupleType) =>
+            r1Final = RecordType(mergeSortedMap(r1Final.fields, tup.toRecord.fields)(_ & _).toList)(noProv)
+            S(that)
           case (N, _) => S(that)
         }) map { b => LhsRefined(S(b), ts, r1) }
```

Comment:
Wait... why is `r1` used here and not `r1Final`? It seems `r1Final` isn't even used!

That's probably a bug, but it seems to have no effect on the current test suite. It would be nice to come up with a test case that makes this incorrect version fail.

Reply:
I didn't wrote this part, so I'm not sure @_@

Reply:
Yeah, I wrote it when we were meeting. Obviously I forgot to use `r1Final` down there, otherwise it makes no sense to have this `var r1Final` (that we update in several places) at all since it does not change anything!

```suggestion
        }) map { b => LhsRefined(S(b), ts, r1Final) }
```